syntax = "proto3";

package yandex.cloud.events.dns;

import "google/rpc/status.proto";
import "google/protobuf/field_mask.proto";
import "yandex/cloud/access/access.proto";
import "yandex/cloud/dns/v1/dns_zone.proto";
import "yandex/cloud/events/common.proto";
import "yandex/cloud/validation.proto";

option go_package = "a.yandex-team.ru/cloud/bitbucket/public-api/yandex/cloud/events/dns;dns";
option java_package = "yandex.cloud.api.events.dns";
option java_outer_classname = "DnsZoneEvents";

message CreateDnsZone {
  Authentication authentication = 1 [(required) = true];
  Authorization authorization = 2 [(required) = true];
  EventMetadata event_metadata = 3 [(required) = true];
  RequestMetadata request_metadata = 4 [(required) = true];
  EventStatus event_status = 5 [(required) = true];
  google.rpc.Status error = 6;
  EventDetails details = 7 [(required) = true];
  RequestParameters request_parameters = 8 [(required) = true];
  events.Response response = 9 [(required) = true];

  message EventDetails {
    string dns_zone_id = 1 [(required) = true];
    string dns_zone_name = 2;
    map<string, string> labels = 3;

    string zone = 4 [(required) = true];
    yandex.cloud.dns.v1.PrivateVisibility private_visibility = 5;
    yandex.cloud.dns.v1.PublicVisibility public_visibility = 6;
  }

  message RequestParameters {
    string folder_id = 1 [(required) = true];
    string name = 2;
    map<string, string> labels = 3;
    string zone = 4 [(required) = true];
    yandex.cloud.dns.v1.PrivateVisibility private_visibility = 5;
    yandex.cloud.dns.v1.PublicVisibility public_visibility = 6;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }
}

message UpdateDnsZone {
  Authentication authentication = 1 [(required) = true];
  Authorization authorization = 2 [(required) = true];
  EventMetadata event_metadata = 3 [(required) = true];
  RequestMetadata request_metadata = 4 [(required) = true];
  EventStatus event_status = 5 [(required) = true];
  google.rpc.Status error = 6;
  EventDetails details = 7 [(required) = true];
  RequestParameters request_parameters = 8 [(required) = true];
  events.Response response = 9 [(required) = true];

  message EventDetails {
    string dns_zone_id = 1 [(required) = true];
    string dns_zone_name = 2;
    map<string, string> labels = 3;

    string zone = 4 [(required) = true];
    yandex.cloud.dns.v1.PrivateVisibility private_visibility = 5;
    yandex.cloud.dns.v1.PublicVisibility public_visibility = 6;
  }

  message RequestParameters {
    string dns_zone_id = 1 [(required) = true];
    google.protobuf.FieldMask update_mask = 2;
    string name = 3;
    map<string, string> labels = 4;
    yandex.cloud.dns.v1.PrivateVisibility private_visibility = 5;
    yandex.cloud.dns.v1.PublicVisibility public_visibility = 6;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }
}

message DeleteDnsZone {
  Authentication authentication = 1 [(required) = true];
  Authorization authorization = 2 [(required) = true];
  EventMetadata event_metadata = 3 [(required) = true];
  RequestMetadata request_metadata = 4 [(required) = true];
  EventStatus event_status = 5 [(required) = true];
  google.rpc.Status error = 6;
  EventDetails details = 7 [(required) = true];
  RequestParameters request_parameters = 8 [(required) = true];
  events.Response response = 9 [(required) = true];

  message EventDetails {
    string dns_zone_id = 1 [(required) = true];
    string dns_zone_name = 2;
    map<string, string> labels = 3;

    string zone = 4 [(required) = true];
    yandex.cloud.dns.v1.PrivateVisibility private_visibility = 5;
    yandex.cloud.dns.v1.PublicVisibility public_visibility = 6;
  }

  message RequestParameters {
    string dns_zone_id = 1 [(required) = true];
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }
}

message UpdateRecordSets {
  Authentication authentication = 1 [(required) = true];
  Authorization authorization = 2 [(required) = true];
  EventMetadata event_metadata = 3 [(required) = true];
  RequestMetadata request_metadata = 4 [(required) = true];
  EventStatus event_status = 5 [(required) = true];
  google.rpc.Status error = 6;
  EventDetails details = 7 [(required) = true];
  RequestParameters request_parameters = 8 [(required) = true];
  events.Response response = 9 [(required) = true];

  message EventDetails {
    string dns_zone_id = 1 [(required) = true];
    string dns_zone_name = 2;
    map<string, string> labels = 3;

    string zone = 4 [(required) = true];
    yandex.cloud.dns.v1.PrivateVisibility private_visibility = 5;
    yandex.cloud.dns.v1.PublicVisibility public_visibility = 6;

    repeated yandex.cloud.dns.v1.RecordSet effective_recordset_deletions = 7;
    repeated yandex.cloud.dns.v1.RecordSet effective_recordset_additions = 8;
  }

  message RequestParameters {
    string dns_zone_id = 1 [(required) = true];
    oneof changes {
      option (exactly_one) = true;
      Update update = 2;
      Upsert upsert = 3;
    }

    message Update {
      repeated yandex.cloud.dns.v1.RecordSet deletions = 1;
      repeated yandex.cloud.dns.v1.RecordSet additions = 2;
    }

    message Upsert {
      repeated yandex.cloud.dns.v1.RecordSet deletions = 2;
      repeated yandex.cloud.dns.v1.RecordSet replacements = 3;
      repeated yandex.cloud.dns.v1.RecordSet merges = 4;
    }
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }
}

message SetDnsZoneAccessBindings {
  Authentication authentication = 1 [(required) = true];
  Authorization authorization = 2 [(required) = true];
  EventMetadata event_metadata = 3 [(required) = true];
  RequestMetadata request_metadata = 4 [(required) = true];
  EventStatus event_status = 5 [(required) = true];
  google.rpc.Status error = 6;
  EventDetails details = 7 [(required) = true];
  RequestParameters request_parameters = 8 [(required) = true];
  events.Response response = 9 [(required) = true];

  message EventDetails {
    string dns_zone_id = 1;
    string dns_zone_name = 2;
    repeated access.AccessBinding access_bindings = 3;
    repeated access.AccessBindingDelta access_binding_deltas = 4;
  }

  message RequestParameters {
    string dns_zone_id = 1;
    repeated access.AccessBinding access_bindings = 2;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }
}

message UpdateDnsZoneAccessBindings {
  Authentication authentication = 1 [(required) = true];
  Authorization authorization = 2 [(required) = true];
  EventMetadata event_metadata = 3 [(required) = true];
  RequestMetadata request_metadata = 4 [(required) = true];
  EventStatus event_status = 5 [(required) = true];
  google.rpc.Status error = 6;
  EventDetails details = 7 [(required) = true];
  RequestParameters request_parameters = 8 [(required) = true];
  events.Response response = 9 [(required) = true];

  message EventDetails {
    string dns_zone_id = 1 [(required) = true];
    string dns_zone_name = 2;
    repeated access.AccessBinding access_bindings = 3;
    repeated access.AccessBindingDelta access_binding_deltas = 4;
  }

  message RequestParameters {
    string dns_zone_id = 1 [(required) = true];
    repeated access.AccessBindingDelta access_binding_deltas = 2;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }
}
